### PR TITLE
Allow for using header X-Forwarded-Host

### PR DESCRIPTION
**Pull request by bkrukowski**
Using `X-Forwarded-Host` header would be useful for proxies

### CODE REVIEW

File: acme/client.go [hunk 1]
```diff
@@ -128,14 +128,14 @@ func (c *Client) SetChallengeProvider(challenge Challenge, p ChallengeProvider)
 //
 // NOTE: This REPLACES any custom HTTP provider previously set by calling
 // c.SetChallengeProvider with the default HTTP challenge provider.
-func (c *Client) SetHTTPAddress(iface string) error {
+func (c *Client) SetHTTPAddress(iface string, allowXForwardedHost bool) error {
```
Comment (mholt) [inline acme/client.go]: :-1: on changing the function signature.

File: acme/http_challenge_server.go [hunk 2]
```diff
@@ -11,17 +11,18 @@ import (
...
+	listener            net.Listener
+	allowXForwardedHost bool
 }
 
 // NewHTTPProviderServer creates a new HTTPProviderServer on the selected interface and port.
 // Setting iface and / or port to an empty string will make the server fall back to
 // the "any" interface and port 80 respectively.
-func NewHTTPProviderServer(iface, port string) *HTTPProviderServer {
-	return &HTTPProviderServer{iface: iface, port: port}
+func NewHTTPProviderServer(iface, port string, allowXForwardedHost bool) *HTTPProviderServer {
```
Comment (mholt) [inline acme/http_challenge_server.go]: Again, let's not change the function signature. :-1:

File: acme/http_challenge_server.go [hunk 3]
```diff
@@ -58,7 +59,15 @@ func (s *HTTPProviderServer) serve(domain, token, keyAuth string) {
...
 	mux := http.NewServeMux()
 	mux.HandleFunc(path, func(w http.ResponseWriter, r *http.Request) {
-		if strings.HasPrefix(r.Host, domain) && r.Method == "GET" {
+		requestHost := r.Host
+		if s.allowXForwardedHost {
+			xForwardedHost := r.Header.Get("x-forwarded-host")
+			if xForwardedHost != "" {
+				requestHost = xForwardedHost
+			}
+		}
```
Comment (mholt) [inline acme/http_challenge_server.go]: Why can't this be done in pre-processing? I'd much rather this kind of thing live outside the acme lib, it seems like kind of a hack.

**Comment by xenolf** [comment]
Hey there! Do I understand this right that this allows for the challenge handler to receive the incoming domain boulder is validating if there is a proxy in between?

**Comment by bkrukowski** [comment]
@xenolf yes, exactly

**Comment by xenolf** [comment]
> @xenolf yes, exactly

Okay. So why not just use the header as a fallback if the `Host` header does not match any expected domain? This would get rid of all the other changes needed and still achieve the goal.
@mholt what do you think?

**Comment by mholt** [comment]
Is there any more risk here of an attacker changing or setting this header value as opposed to changing the Host header?

**Comment by ldez** [comment]
It's highly insecure because `X-Forwarded-Host` can be spoof.

**Comment by ldez** [comment]
It's time to close and decline this PR.

Thanks.

**Pull request closed without merging**